core/src/price_estimation/clients/apiclient.rs

```diff
@@ -1,79 +1,88 @@
+    /// Returns the price of one unit of `from` expressed in `to`.
+    /// For example `get_price("ETH", "DAI")` is ~220.
+    fn get_price<'a>(&'a self, from: &Self::Token, to: &Self::Token) -> BoxFuture<'a, Result<f64>>;
+}
+
+pub trait HttpConnecting: Sized {
+    /// Creates a new HTTP interface for connecting to the API
+    fn bind(http_factory: &HttpFactory) -> Result<Self>;
+}
+
+struct Tokens<T: Api> {
     // Maps uppercase Token::symbol to Token.
     // This is cached in the struct because we don't expect it to change often.
-    tokens: HashMap<String, api::Token>,
-    stable_coin: api::Token,
+    tokens: HashMap<String, T::Token>,
+    stable_coin: T::Token,
 }
 
-pub struct DexagClient<Api> {
-    api: Api,
+pub struct Client<T: Api> {
+    api: T,
     /// Lazily retrieved the first time it is needed when `get_prices` is
     /// called. We don't want to use the network in `new`.
-    api_tokens: Mutex<Option<ApiTokens>>,
+    api_tokens: Mutex<Option<Tokens<T>>>,
     tokens: TokenData,
 }
 
-impl DexagClient<DexagHttpApi> {
-    /// Create a DexagClient using DexagHttpApi as the api implementation.
+impl<T> Client<T>
+where
+    T: Api + HttpConnecting,
+{
+    /// Create a Client using the api implementation from HttpConnecting.
     pub fn new(http_factory: &HttpFactory, tokens: TokenData) -> Result<Self> {
-        let api = DexagHttpApi::new(http_factory)?;
+        let api = T::bind(http_factory)?;
         Ok(Self::with_api_and_tokens(api, tokens))
     }
-}
 
-impl<Api> DexagClient<Api>
-where
-    Api: DexagApi,
-{
-    pub fn with_api_and_tokens(api: Api, tokens: TokenData) -> Self {
+    pub fn with_api_and_tokens(api: T, tokens: TokenData) -> Self {
         Self {
             api,
             api_tokens: Mutex::new(None),
             tokens,
         }
     }
 
-    async fn create_api_tokens(&self) -> Result<ApiTokens> {
+    async fn create_api_tokens(&self) -> Result<Tokens<T>> {
         let tokens = self.api.get_token_list().await?;
-        let mut tokens: HashMap<String, api::Token> = tokens
+        let mut tokens: HashMap<String, T::Token> = tokens
             .into_iter()
-            .map(|token| (token.symbol.to_uppercase(), token))
+            .map(|token| (token.symbol().to_uppercase(), token))
             .collect();
 
-        // We need to return prices in OWL but Dexag does not track it. OWL
-        // tracks USD so we use another stable coin as an approximate
-        // USD price.
+        // We need to return prices in OWL but some price sources do not track it.
+        // OWL tracks USD so we use another stable coin as an approximate USD price.
         const STABLE_COIN: &str = "DAI";
-        let stable_coin = tokens.remove(STABLE_COIN).ok_or_else(|| {
-            anyhow!(
-                "dexag exchange does not track our stable coin {}",
-                STABLE_COIN
-            )
-        })?;
+        let stable_coin = tokens
+            .remove(STABLE_COIN)
```

Comment:
Stable coin should probably be a constructor argument so that we can use DAI for Dex.Ag and OWL for 1inch

Reply:
Agreed that `STABLE_COIN` should not be constant. It's now returned by a method of the `Api` trait, in this way it's easily customizable when creating a new Api implementation.